[docs-infra] Fix search ranking when no productId

oliviertassinari opened the pull request:
I broke this when I refactored the productId logic. The problem is that we were setting `productId:null` as an optional filter in the homepage. Now, Algolia is configured with

```js
      ranking: [
        "words",
        "filters",
        "typo",
        "attribute",
        "proximity",
        "exact",
        "custom",
      ],
```

https://www.algolia.com/doc/guides/managing-results/relevance-overview/in-depth/ranking-criteria/#filters

so it would rank first all the pages without a productId, which is right now the MUI X migration page as we have to refactor a bit these pages to get the productId from the markdown.

mnajdova commented:
This solves the migration pages appearing first 👍 But still, do we really want the Toolpad "text field" to appear first on the home page search? I feel like we should rank the results on the home page based on the popularity of the product - meaning Core, X, Toolpad. We can address this separately.

oliviertassinari commented:
> But still, do we really want the Toolpad "text field" to appear first on the home page search? I feel like we should rank the results on the home page based on the popularity of the product - meaning Core, X, Toolpad. We can address this separately.

@mnajdova Yeah, but this would be another problem, different from the one this PR fixes. In this case, Toolpad ranks first because it has an exact word match. Then if you search for "text field", Toolpad still ranks first because it comes first when you sort the URLs in an alphabetical descending order.

siriwatknp reviewed (approved):
Some comments in the code would be nice, e.g. where to check the filter configuration of our site.

oliviertassinari commented:
@siriwatknp I think we could document this in https://www.notion.so/mui-org/Docs-of-the-docs-infra-7ee89042350745f29803413b28d5b60f?pvs=4#15e70fafa7754b2aa6be5bfca82978ef. I have added a bit of content.

I have added a `pageRank`.

```jsx
        // % from https://docs.google.com/spreadsheets/d/1x9uTSMlu4xug2SfHrvr5amZ2HokYBoQaSozHmRRdVqw/edit#gid=907743010
        const baselinePopulartyProductId = {
          "material-ui": 83,
          "mui-x": 12,
          "system": 4,
          "base-ui": 2,
          "joy-ui": 2,
          toolpad: 1,
        };

        const pageRank =
          // @ts-ignore
          baselinePopulartyProductId[productId] ||
          // @ts-ignore
          baselinePopulartyProductId[productCategoryId] ||
          0;
```

To solve the ranking between products on the homepage.

<img width="751" alt="Screenshot 2023-07-26 at 14 50 45" src="https://github.com/mui/material-ui/assets/3165635/5d447511-76c5-4b84-9d85-8214977d02a4">

I have also used the popularity for the ranking between component page

```jsx
        } else if (url.pathname.startsWith("/material-ui/api/")) {
          popularity = -1;
```

to solve the ranking between API pages and demo pages:

<img width="751" alt="Screenshot 2023-07-26 at 14 50 45" src="https://github.com/mui/material-ui/assets/3165635/6c53225d-ffb5-4134-a7f0-b6c6a2d9a274">

However, there are still a lot of broken things. For instance, the way the grouping it down by Algolia is fucked up. It relies on the `lvl0` field: https://github.com/algolia/docsearch/blob/2aef60fcfdecdae0cce2897b6213efa62f7dee80/packages/docsearch-react/src/utils/removeHighlightTags.ts#L19 but this is the same between Material UI, Joy UI, Base UI, so groups are nor correctly formed 🙈

oliviertassinari commented:
Anyway, this is getting way outside my initial objective, this PR is a quick win, fixing an obvious mistake 😁
